server/src/main/java/org/elasticsearch/search/vectors/VectorData.java

```diff
@@ -0,0 +1,168 @@
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HexFormat;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.common.Strings.format;
+
+public record VectorData(float[] floatVector, byte[] byteVector) implements Writeable, ToXContentFragment {
+
+    public static final String XCONTENT_PARAM_NAME = "field";
+    private static final String DEFAULT_XCONTENT_NAME = "vector";
+
+    private VectorData(float[] floatVector) {
+        this(floatVector, null);
+    }
+
+    private VectorData(byte[] byteVector) {
+        this(null, byteVector);
+    }
+
+    public VectorData(StreamInput in) throws IOException {
+        this(in.readOptionalFloatArray(), in.readOptionalByteArray());
+    }
+
+    public VectorData {
+        if (false == (floatVector == null ^ byteVector == null)) {
+            throw new IllegalArgumentException("please supply exactly either a float or a byte vector");
+        }
+    }
+
+    public byte[] asByteVector() {
+        if (byteVector != null) {
+            return byteVector;
+        }
+        DenseVectorFieldMapper.ElementType.BYTE.checkVectorBounds(floatVector);
+        byte[] vec = new byte[floatVector.length];
+        for (int i = 0; i < floatVector.length; i++) {
+            vec[i] = (byte) floatVector[i];
+        }
+        return vec;
+    }
+
+    // Explicit byte parsing is only supported when operation on hex-encoded byte vectors
+    // which are defined only for ElementType.BYTE, so we should throw if we request to convert it to float
+    public float[] asFloatVector() {
+        return asFloatVector(true);
+    }
+
+    public float[] asFloatVector(boolean failIfByte) {
+        if (byteVector != null) {
+            if (failIfByte) {
+                throw new UnsupportedOperationException("cannot convert to float, as we're explicitly using a byte vector");
+            } else {
+                float[] vec = new float[byteVector.length];
+                for (int i = 0; i < byteVector.length; i++) {
+                    vec[i] = byteVector[i];
+                }
+                return vec;
+            }
+        }
+        return floatVector;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalFloatArray(floatVector);
+        out.writeOptionalByteArray(byteVector);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        if (floatVector != null) {
+            builder.array(params.param(XCONTENT_PARAM_NAME, DEFAULT_XCONTENT_NAME), floatVector);
+        } else {
+            builder.array(params.param(XCONTENT_PARAM_NAME, DEFAULT_XCONTENT_NAME), byteVector);
+        }
+        return builder;
+    }
```

Comment:
It would be much simpler if this would only write the non-null array without a field name. Then the containing objects would do `builder.field(fieldName, vectorData);` which will get written as `fieldName: [1, 2, 3...]`

I think the use of XContentParams here, while interesting, is complex.

Reply:
Yeap I agree with your point. Tbh that's how I had it initially, but some serialization tests were failing with the following, so I resorted to making use of the more complex approach using `XContentParams`
```
...
Caused by: com.fasterxml.jackson.core.JsonGenerationException: Can not start an array, expecting field name
	at com.fasterxml.jackson.core.JsonGenerator._reportError(JsonGenerator.java:2849)
	at com.fasterxml.jackson.dataformat.yaml.YAMLGenerator._verifyValueWrite(YAMLGenerator.java:916)
	at com.fasterxml.jackson.dataformat.yaml.YAMLGenerator.writeStartArray(YAMLGenerator.java:586)
	at org.elasticsearch.xcontent.provider.json.JsonXContentGenerator.writeStartArray(JsonXContentGenerator.java:169)
```
I ended up re-writing the tests using mocks instead of concrete instances for the validation of `VectorData#toXContent`, but failed to change this one back. 

 ++ for the change, will update it.